### PR TITLE
Replace `_DataModuleWrapper` with `__new__` [1/2]

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

We now rely on `__new__` to wrap and track the `DataHooks` calls.

## Fixes:

This fixes a bug where the `_DataModuleWrapper` implementation was adding as many wrapped layers as the number of DataModule subclasses.
This is not a problem if a wrapped layer does not impact any further wrapper layers which is no longer the case after #7238 (blocked by this PR).
And even without #7238, we should not wrap these hooks multiple times.

To compare the behaviors, see the following minimal examples:

#### Current master

```python
import functools
import pickle


class Metaclass(type):
    def __init__(cls, *args, **kwargs):
        super().__init__(*args, **kwargs)
        cls.has_wrapped = False

    def __call__(cls, *args, **kwargs):
        if not cls.has_wrapped:
            cls.thing = cls.wrapper(cls.thing)
            cls.has_wrapped = True
        obj = type.__call__(cls, *args, **kwargs)
        return obj

    @staticmethod
    def wrapper(fn):
        @functools.wraps(fn)
        def wrapped_fn(*args, **kwargs):
            print(f"Wrapped {fn.__name__}. args: {args}, kwargs: {kwargs}")
            return fn(*args, **kwargs)
        return wrapped_fn


class Foo(metaclass=Metaclass):
    def thing(self):
        print(f"{self.__class__.__name__}.thing()")

class Bar(Foo): ...
class FooBar(Bar): ...


f = Foo()
b = Bar()
fb = FooBar()

pickle.dumps(f)

f.thing()
b.thing()
fb.thing()
"""
Wrapped thing. args: (<__main__.Foo object at 0x10f6ef940>,), kwargs: {}
Foo.thing()
Wrapped thing. args: (<__main__.Bar object at 0x10f6ef820>,), kwargs: {}
Wrapped thing. args: (<__main__.Bar object at 0x10f6ef820>,), kwargs: {}
Bar.thing()
Wrapped thing. args: (<__main__.FooBar object at 0x10f6ef8e0>,), kwargs: {}
Wrapped thing. args: (<__main__.FooBar object at 0x10f6ef8e0>,), kwargs: {}
Wrapped thing. args: (<__main__.FooBar object at 0x10f6ef8e0>,), kwargs: {}
FooBar.thing()
"""
```

#### This PR

```python
import functools
import pickle


class Foo:
    def __new__(cls, *args, **kwargs):
        obj = super(Foo, cls).__new__(cls)
        obj.__args = args
        obj.__kwargs = kwargs
        obj.thing = cls.wrapper(obj, obj.thing)
        return obj

    def __getstate__(self):
        # avoids _pickle.PicklingError: Can't pickle <...>: it's not the same object as <...>
        d = self.__dict__.copy()
        del d["thing"]
        return d

    @staticmethod
    def wrapper(obj, fn):
        @functools.wraps(fn)
        def wrapped_fn(*args, **kwargs):
            print(f"Wrapped {obj}.{fn.__name__}. args: {args}, kwargs: {kwargs}")
            return fn(*args, **kwargs)
        return wrapped_fn

    def thing(self):
        print(f"{self.__class__.__name__}.thing()")

class Bar(Foo): ...
class FooBar(Bar): ...


f = Foo()
b = Bar()
fb = FooBar()

# needs `__getstate__`
pickle.dumps(f)

f.thing()
b.thing()
fb.thing()
"""
Wrapped <__main__.Foo object at 0x1075bb970>.thing. args: (), kwargs: {}
Foo.thing()
Wrapped <__main__.Bar object at 0x1075bb910>.thing. args: (), kwargs: {}
Bar.thing()
Wrapped <__main__.FooBar object at 0x1074195b0>.thing. args: (), kwargs: {}
FooBar.thing()
"""
```

# Question:

How do I even test this automatically?

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/core/datamodule.py [hunk 1]
```diff
@@ -398,3 +325,62 @@ def test_dataloader():
...
+        Args:
+            obj: Object whose function will be tracked
+            fn: Function that will be tracked to see if it has been called.
+
+        Returns:
+            Decorated function that tracks its call status and saves it to private attrs in its obj instance.
+        """
+
+        @functools.wraps(fn)
+        def wrapped_fn(*args: str, **kwargs: Optional[str]) -> callable:
```
Comment (carmocca) [inline pytorch_lightning/core/datamodule.py:359]:
```suggestion
        def wrapped_fn(*args: str, **kwargs: Optional[str]) -> Any:
```

**Review by ananthsub — COMMENTED** [review]
Would it be simpler to have explicit properties and setters on the datamodule which the trainer can set after calling the hooks? 

I also am not sure if the current properties are correct for setup and teardown if one re-uses the same module across consecutive calls. 

What's expected for this?  

trainer.fit(model, dm)
trainer.fit(model, dm) 

Should the second call skip calling the datamodule hooks?

**Comment by carmocca** [comment]
> Would it be simpler to have explicit properties and setters on the datamodule which the trainer can set after calling the hooks?

This is the case already. But instead of setting them manually by the trainer, it is done automatically using these "hacks".

> Should the second call skip calling the datamodule hooks?

This is also the case already.

It is not if the user does the call manually, and for that I have #7238

**Comment by ananthsub** [comment]
Actually why do we need these setters at all? Couldn't we push the responsibility to the datamodule on whether it needs to download data again? Then we can simplify the trainer and avoid all the hook wrappers here

**Comment by carmocca** [comment]
If I had designed this from the beginning I would have done one of the 2 following options instead of what we have:

1. Move the responsability to the user so they check if the files are already downloaded and splits already created (if they want)
2. Ask users to call `super().setup(stage)` and in the base class we set these `has_setup_stage` checks.

But this is not the current design and I don't think we can safely switch the behaviour.

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -398,3 +325,62 @@ def test_dataloader():
         if test_dataset is not None:
             datamodule.test_dataloader = test_dataloader
         return datamodule
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> 'LightningDataModule':
+        obj = super(LightningDataModule, cls).__new__(cls)
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:330]: do you need to specify the arguments to `super` here? Couldn't you just call `super()`?

File: pytorch_lightning/core/datamodule.py [hunk 3]
```diff
@@ -398,3 +325,62 @@ def test_dataloader():
...
+        obj.__args = args
+        obj.__kwargs = kwargs
+        # track `DataHooks` calls and run `prepare_data` only on rank zero
+        obj.prepare_data = cls._track_data_hook_calls(obj, rank_zero_only(obj.prepare_data))
+        obj.setup = cls._track_data_hook_calls(obj, obj.setup)
+        obj.teardown = cls._track_data_hook_calls(obj, obj.teardown)
+        return obj
+
+    @staticmethod
+    def _track_data_hook_calls(obj: 'LightningDataModule', fn: callable) -> callable:
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:341]: Can we easily move the function here? I doubt that there will be user code, but you never know. So maybe we should have an alias with deprecation warning to ensure backwards compatibility?

File: pytorch_lightning/core/datamodule.py [hunk 4]
```diff
@@ -398,3 +325,62 @@ def test_dataloader():
...
+            elif name == "prepare_data":
+                obj._has_prepared_data = True
+
+            return fn(*args, **kwargs)
+
+        return wrapped_fn
+
+    def __reduce__(self) -> Tuple[type, tuple, dict]:
+        # avoids _pickle.PicklingError: Can't pickle <...>: it's not the same object as <...>
+        return self.__class__, self.__args, self.__kwargs
```
Comment (justusschock) [inline pytorch_lightning/core/datamodule.py:386]: can't we maybe pickle the `__dict__` instead?

**Review by justusschock — COMMENTED** [review]
I left some comments. It looks good to me. However, I don't enough about these things for being confident enough to approve.

File: pytorch_lightning/core/datamodule.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:330]: Doesn't seem like it:

```suggestion
        obj = super().__new__(cls)
```

File: pytorch_lightning/core/datamodule.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:341]: GitHub code search does not show a single occurrence:

https://github.com/search?q=track_data_hook_calls&type=code

I have no idea why this wasn't protected initially, it is a very internal piece of code.

An alias would not work as it now takes an `obj` instance too. To keep BC I would have to keep the old copy with `pragma: no-cover`.

I don't think it's worth it.

**Review by kaushikb11 — APPROVED** [review]

**Comment by ananthsub** [comment]
> But this is not the current design and I don't think we can safely switch the behaviour.

Do you think we can add warnings in the trainer if the `has_{setup/teardown}_*` properties are True? And then we can phase out these checks in v1.5?

**Comment by carmocca** [comment]
> > But this is not the current design and I don't think we can safely switch the behaviour.
> 
> Do you think we can add warnings in the trainer if the `has_{setup/teardown}_*` properties are True? And then we can phase out these checks in v1.5?

Do you mean if we go with option (1)?:
> Move the responsability to the user so they check if the files are already downloaded and splits already created (if they want)

Do you want to open an RFC or discussion about this?

**Review by Borda — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Pull request merged**